### ISSUE TITLE
Result of BaseGraph on computer and espdl model on ESP32-S3 are not equal

### ISSUE THREAD

**Issue by CzakelMM** [espressif/esp-dl#246]
### Checklist

- [x] Checked the issue tracker for similar issues to ensure this is not a duplicate.
- [x] Provided a clear description of your suggestion.
- [x] Included any relevant context or examples.

#### Description

I have build model base on MobileNetV2 for custom image classification.
This is modified model which has fallowing parameters:
- input - 1x256x64
- output - 1x5
- width_mult - 0.8

Model was developed with PyTorch and quantized with exactly the same parameters like in your example (file  `examples/tutorial/how_to_quantize_model/quantize_mobilenetv2/quantize_torch_model.py`)

Results of quantized model  calculated on computer with BaseGraph (return from function `espdl_quantize_torch`) are correct.

But when i uploaded ESPDL model to ESP32-S3 partition and run model in firmware, I have notice that model results calculated on computer with BaseGraph and by the ESPDL model on ESP32-S3 are not equal. ESPDL model returned worse results.

When i run your example (folder `examples/tutorial/how_to_quantize_model/quantize_mobilenetv2`)  with standard MobileNetv2 (input 3x224x224, output 1x1000) with images from imagenet dataset, results of BaseGraph on computer and result of ESPDL model on ESP32-S3 were equal.

#### Attachment
I have attached ZIP file with:
- image.raw - already transformed image (normalized, float range <1,0>) , input to model 
- model.espdl - ESPDL model, result from function `espdl_quantize_torch`  
- model.graph - BaseGraph returned from function `espdl_quantize_torch` 

#### Code

Python code to get result from BaseGraph on the comptuer
```
import torch
import numpy as np
from ppq.api import load_graph
from ppq.core import NetworkFramework
from ppq.executor.torch import TorchExecutor

img_path = "image.raw"
graph_path = "model.graph"

def img_to_tensor(img_path):
    with open(img_path, 'rb') as f:
        data = f.read()

    numpy_array = np.frombuffer(data, dtype=np.float32).reshape((1, 64, 256)).copy()
    return torch.from_numpy(numpy_array).unsqueeze(0)

input = img_to_tensor(img_path)

graph = load_graph(graph_path, from_framework=NetworkFramework.NATIVE)
model = TorchExecutor(graph=graph, device="cpu")

y_preds = model(input)[0]
y_probs = torch.softmax(y_preds, dim=1).squeeze(0)

print("Softmax Output:")
for i, prob in enumerate(y_probs):
    print(f"[{i}] {prob:.6f}")
```
Received results from BaseGraph:
```
Softmax Output:
[0] 0.787685
[1] 0.009915
[2] 0.128586
[3] 0.068827
[4] 0.004986
```

Cpp code of ESP32-S3 firmware, which load espdl model from partition and image from embed file (exact the same like on computer) and calculate model result:
```
#include "dl_model_base.hpp"
#include "dl_module_softmax.hpp"

extern const float img_start[] asm("_binary_image_raw_start");
extern const float img_end[] asm("_binary_image_raw_end");

extern "C" void app_main(void)
{
    dl::Model *model = new dl::Model("model", fbs::MODEL_LOCATION_IN_FLASH_PARTITION);

    std::map<std::string, dl::TensorBase *> model_inputs = model->get_inputs();
    dl::TensorBase *model_input = model_inputs.begin()->second;
    std::map<std::string, dl::TensorBase *> model_outputs = model->get_outputs();
    dl::TensorBase *model_output = model_outputs.begin()->second;

    dl::TensorBase *input_tensor = new dl::TensorBase(
        model_input->get_shape(), 
        img_start,
        0,
        dl::DATA_TYPE_FLOAT
    );
    model_input->assign(input_tensor);

    model->run();

    dl::TensorBase *output_tensor = new dl::TensorBase(model_output->shape, nullptr, 0, dl::DATA_TYPE_FLOAT);

    dl::module::Softmax* softmax_module = new dl::module::Softmax(nullptr, -1, dl::MODULE_NON_INPLACE, dl::QUANT_TYPE_SYMM_8BIT);
    softmax_module->run(model_output, output_tensor);

    printf("Softmax Output:\n");
    int output_size = model_output->shape[1];
    for (int i = 0; i < output_size; ++i) {
        printf("[%d]: %.6f\n", i, output_tensor->get_element<float>(i));
    }

    delete input_tensor;
    delete output_tensor;
    delete softmax_module;
    delete model;
}
```
Received results from ESPDL model:
```
Softmax Output:
[0]: 0.612001
[1]: 0.034527
[2]: 0.025260
[3]: 0.039124
[4]: 0.289089
```

#### Questions

1. Do BaseGraph on computer and ESPDL model on ESP32 should return exactly the same results?
2. Do i correctly input image data into the model in firmware?
3. Is this a result of incorrect quantization?

[model_image.zip](https://github.com/user-attachments/files/20661112/model_image.zip)

**Comment by sun-xiangyu** [comment]
> * Do BaseGraph on computer and ESPDL model on ESP32 should return exactly the same results?
> * Do i correctly input image data into the model in firmware?
> * Is this a result of incorrect quantization?

The results from the computer and the ESP32 should be the same for int8 quantization. I'm not sure whether you've ported the ESP32 and computer implementations correctly.
I recommend you first use the built-in testing functionality of ​​esp-ppq​​. 
1. please modify the export parameters of ​​esp-ppq​​ to directly save the ​​inputs​​ and ​​outputs​​ from the Python test into the ​​espdl​​ model, like this https://github.com/espressif/esp-dl/blob/578ca7ada61dadc2ff769c46a3e7b689cc37e3ab/tools/ops_test/gen_test_cases.py#L135
2. then call [model->test()](https://github.com/espressif/esp-dl/blob/578ca7ada61dadc2ff769c46a3e7b689cc37e3ab/examples/tutorial/how_to_load_test_profile_model/model_in_flash_partition/main/app_main.cpp#L12) to perform the testing.

**Comment by CzakelMM** [comment]
I have figure out the problem. It was stupid mistake.

Warning for anyone who design model where input image shape isn't square. In default MobileNetV2 input image shape is 3x224x224 but in my case input image is rectangular 1x256x64. 

I have input incorrect shape to argument `input_shape` in function `espdl_quantize_torch`. I have input shape `1x1x256x64` but it should be `1x1x64x256`. PyTorch tensor (and also numpy array) hold image in format [color, row, column]. When incorrect shape is inputted to function `espdl_quantize_torch` model is still quantize successfully, but ESPDL model wouldn't work correctly on ESP32

Issue to close.
